### PR TITLE
refactor: Convert fields to ES6 classes

### PR DESCRIPTION
**Pull request by BeksOmega**
## The basics

- [X] I branched from develop
- [X] My pull request is against develop
- [X] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)

### Resolves

Work on #5860 

### Proposed Changes

Remove setting value (and therefore triggering validation) from the base field class. Each individual subclass will now trigger these things, so we can make sure necessary properties are defined before validation is triggered.

Use a `Field.SENTINEL` value to make sure that "intermediate" classes skip setting the value and triggering validation when necessary.

#### Behavior before change

Before it was possible to set a `<Subclass>.prototype.DEFAULT_VALUE` property which the base `Field` class would set as the default value in case validation were to fail.

#### Behavior after change

As far as I can think, there's no way to make `DEFAULT_VALUE` available to the base `Field` class, so Devs will need to set a default themselves instead.

### Reason for Changes

Making things ES6 while still making sure validation of values passed to constructors happens.

### Documentation

Might want to update the [constructor section of the fields doc](https://developers.google.com/blockly/guides/create-custom-blocks/fields/customizing-fields/creating#implementing_a_constructor)? But it looks like it was never exactly correct to begin with.

### Additional Information

There are still lots of broken things and fields to convert, this is just a draft PR so people can take a look at the proposed solution.

### CODE REVIEW

File: core/field.js
Comment (BeksOmega) [inline core/field.js:535]: I think this should work to check for overrides?

**Review by cpcallen — CHANGES_REQUESTED** [review]
Nice job.  What a monster!  I have comments:

* I think with such a large PR it would have been helpful to have presented it as a series of commits, each tackling one particular part of the task, that could be reviewed in sequence.  I started off trying to review the commits one at a time but quickly realised that some of your later PRs undid changes in the earlier ones; unfortunately reviewing all changes was also sub-optimal because clang-format changes were mixed in with functional ones.  (In hindsight I'd have suggested you hold off on running clang-format until everything else had been reviewed and approved.)

* I've tried not to pick on you too much for stuff that's not actually your fault, but apologies for not being entirely successful at that.

* You wrote:
  > Before it was possible to set a `<Subclass>.prototype.DEFAULT_VALUE` property which the base `Field` class would set as the default value in case validation were to fail.

  > As far as I can think, there's no way to make `DEFAULT_VALUE` available to the base `Field` class, so Devs will need to set a default themselves instead.

  Amazingly enough, there is, thanks to [`new.target`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/new.target)!  The constructor can just read `new.target.prototype.DEFAULT_VALUE`.

* I'm wondering about the memory usage of moving all these properties from the prototype object(s) to the instances.  Might it be a good idea to do a quick sanity check of before/after, just to make sure it hasn't caused memory usage to explode when loading a program with lots of blocks?  (Though it could conceivably _save_ memory in V8 by not requiring so many shadow classes…)

* There are a lot of comments that mention "`field`".  Since the word field has several meanings, I suggest using `Field` when referring to the class _or instances of it_.

Specifically on the topic of the sentinel:

1. I note that, so far in the code I have seen, when we pass `SENTINEL` it is the only parameter.  I'm not sure we want it to cause any other parameters passed to be ignored (see comment inline below).
0. But if we do then we could eliminate having a sentinel entirely by having the `Field` constructor (and its subclass constructors) skip initialisation if not passed any arguments.
   * The easy way to do this is to check `arguments.length` and skip the setup code if it was `0`.
   * Alternatively this could be implemented by supplying `SENTINEL` as the default for `value`—note that it can now be unexported because callers never need it (though subclasses might need their own equivalent):a
     ```JS
     const SENTINEL = {};
     class Field {
       constructor(value = SENTINEL, /* ... */) {
         // ...
     ```
     (This might also be faster than using `arguments.length`: there used to be a cost to even mentioning `arguments` as the engine would have to instantiate an `Arguments` object; not sure if this gets optimised away these days or not.)
   Either way is technically a breaking change for anyone who was writing `new Field()` as shorthand for `new Field(undefined)`, but I think that is probably acceptable.
0. In any case, if we continue to use a sentinel then it needs a better name, that more clearly explains what it does—perhaps something like `NO_VALUE_YET` or `SKIP_SETUP` or even just `DEFERRED`.
0. I also think it would be helpful to make it a (singleton) instance of its own named type, so that type declarations can allow, say, `...|!Field.Sentinel` rather than `...|Object`:
   ```JS
   Field.Sentinel = class {};
   field.BETTER_NAME_FOR_SINGLETON = new Field.Sentinel();
   ```

**Comment by cpcallen** [comment]
Oh, one thing I forgot: I think we should move `DEFAULT_VALUE` from `.prototype` to the constructor (as a static property), though for backwards compatibility the constructor could check `value` then `new.target.prototype.DEFAULT_VALUE` (if it exists because of legacy developer code), then `new.target.DEFAULT_VALUE` (which will hopefully always exist because we can define it).

File: core/field.js
Comment (BeksOmega) [reply core/field.js:90]: Yeah that would be sweet! But it would definitely have to be a bigger refactor because names are actually optional (most FieldLabels don't have names for example). And names are actually passed to the Input rather than the Field itself via the [appendField](https://github.com/google/blockly/blob/e21a8fcd0cf16fd26b589d6fbf42d469ca57e863/core/input.js#L103) function.

As a bit of trivia, Field names should always be unique within a block, because they are used for lookup, but lookup is done using a [fun nested loop](https://github.com/google/blockly/blob/e21a8fcd0cf16fd26b589d6fbf42d469ca57e863/core/block.js#L1071) rather than a dictionary.

File: core/field.js
Comment (BeksOmega) [reply core/field.js:568]: I mean it is weird definitely, but idk what an alternative would be. [isClickable](https://github.com/google/blockly/blob/e21a8fcd0cf16fd26b589d6fbf42d469ca57e863/core/field.js#L705) is only used to determine if `showEditor_` should be called when the field is clicked (so we definitely want `showEditor_` to be overridden, since the base definition does nothing). I'll admit that this is also weird.

Theoretically it would be nice to refactor this, but `isClickable` is public, so I don't think we can really change its definition at this point.

File: core/field.js
Comment (BeksOmega) [reply core/field.js:709]: Yeah :/ `showEditor_` was the thing that originally existed, which built-in Fields were overriding. (This was back when Blockly didn't really use package or protected annotations) so we're stuck with this.

File: core/field_dropdown.js
Comment (BeksOmega) [reply core/field_dropdown.js:162]: Answer to 1: That type annotation is super wrong, it should be `!Array<*>`, because `getOptions` is `!Array<!Array>

Answer to 2: I thought [override](https://github.com/google/closure-compiler/wiki/Annotating-JavaScript-for-the-Closure-Compiler#override) triggered annotation inheritance. Do I not need to do that?

Answers to 3 & 4:
This is just trying to follow the patterns of other Fields, which set a default value before calling `setValue` incase validation fails. The interesting thing about dropdown fields is that their initial value is guaranteed to be valid, because (1) the default option is the first option in the menu, and (2) a value is valid if it exists in the menu.

So this should just call `this.setValue(this.selectedOption_[1])` which I will do.

File: core/field_label_serializable.js
Comment (BeksOmega) [reply core/field_label_serializable.js:43]: Duuuude! I didn't know that javascript had a nullish coalescing operator! This is so exciting!

File: core/field_number.js
Comment (BeksOmega) [reply core/field_number.js:94]: Yeee this is another backwards compatibility thing. The config param didn't always exist. But we didn't want constructors to keep growing and growing, so we wanted to add a config. And we ended up with this weird half-half thing.

File: core/field.js
Comment (cpcallen) [reply core/field.js:568]: OK, that sounds reasonable enough.  I guess the "pure" design would be for this method just to return false, and subclasses that override `showEditor_` would also have to override `isClickable` to return true when applicable, but I admit that the current implementation seems more helpful.

File: core/field.js
Comment (cpcallen) [reply core/field.js]: Yes.

File: core/field_dropdown.js
Comment (cpcallen) [reply core/field_dropdown.js:162]: > I thought [override](https://github.com/google/closure-compiler/wiki/Annotating-JavaScript-for-the-Closure-Compiler#override) triggered annotation inheritance. Do I not need to do that?

For methods that is recommended, but here I think you are just changing the value of the `.value_` property declared in the superclass, rather than overriding anything.

File: core/field_label_serializable.js
Comment (cpcallen) [reply core/field_label_serializable.js:43]: It does now.  (Though we may need to adjust eslint and Closure Compiler settings…)

**Comment by BeksOmega** [comment]
> I'm wondering about the memory usage of moving all these properties from the prototype object(s) to the instances. Might it be a good idea to do a quick sanity check of before/after, just to make sure it hasn't caused memory usage to explode when loading a program with lots of blocks? (Though it could conceivably save memory in V8 by not requiring so many shadow classes…)

Ok I did some profiling, and there's definitely an increase, but I don't think it is incredibly significant. The two allocation timelines I took are linked below. For these recordings I first ran the garbage collector, and then ran spaghetti.

[Old fields allocation](https://drive.google.com/file/d/1WEdGDSRpxaYNQEd1Y5f938IJIAxO-qag/view).
[New fields allocation](https://drive.google.com/file/d/12BSISgiPEldJH4Mk30-TxHSOvYoqAc_y/view).

Looking at `FieldNumber` as an example, you can see that the shallow size increased by 1.4x and the retained size increased by 1.1x. However, the memory percentage allocated to FieldNumber is still ~1%.

**Comment by BeksOmega** [comment]
> Amazingly enough, there is, thanks to [new.target](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/new.target)! The constructor can just read new.target.prototype.DEFAULT_VALUE.

Coolio! That seems to work. The only issue is I'm getting a debug build error. Where does the closure config live?
```
core/field.js:98:39: ERROR - [JSC_INEXISTENT_PROPERTY] Property DEFAULT_VALUE never defined on Function.prototype
  98|     this.value_ = new.target.prototype.DEFAULT_VALUE;
```

File: core/field.js
Comment (BeksOmega) [reply core/field.js:181]: Wait, looking at this again, wouldn't that make `DEFAULT_VALUE` a property on the class constructor rather than the prototype? I don't think we want that.

File: core/field.js
Comment (BeksOmega) [reply core/field.js:266]: So wrt to the config, I think we definitely *don't* want to apply that because it will trigger the same problems that setting the value does. Namely properties getting set by subclasses before they have been initialized in that subclass' constructor.

Then wrt to the validator, I don't think we want to apply that until after we set the initial value, because we don't want it to be triggered until after the field is attached to a block. This is because many external devs create validators that rely on other state in the workspace. So they expect to have a reference to `this.getSourceBlock()` whenever their validator is called. If we break that implicit promise, it will probably be pretty bad.

Now: wrt the question of how to communicate that we're ignoring all of the parameters, I am not sure.

**Comment by BeksOmega** [comment]
> Specifically on the topic of the sentinel:
>
> 1. I note that, so far in the code I have seen, when we pass `SENTINEL` it is the only parameter.  I'm not sure we want it to cause any other parameters passed to be ignored (see comment inline below).

Responded inline above.

> 0. But if we do then we could eliminate having a sentinel entirely by having the `Field` constructor (and its subclass constructors) skip initialisation if not passed any arguments.
>    * The easy way to do this is to check `arguments.length` and skip the setup code if it was `0`.
>    * Alternatively this could be implemented by supplying `SENTINEL` as the default for `value`—note that it can now be unexported because callers never need it (though subclasses might need their own equivalent):a
     ```JS
     const SENTINEL = {};
     class Field {
       constructor(value = SENTINEL, /* ... */) {
         // ...
     ```
     (This might also be faster than using `arguments.length`: there used to be a cost to even mentioning `arguments` as the engine would have to instantiate an `Arguments` object; not sure if this gets optimised away these days or not.)
   Either way is technically a breaking change for anyone who was writing `new Field()` as shorthand for `new Field(undefined)`, but I think that is probably acceptable.

As discussed ofline last week, Currently it is totally legal to do something like `new FieldTextInput()` or `new FieldNumber()` if you want a plain old Field with the default value. I don't think we want to break this right now :/

> 0. In any case, if we continue to use a sentinel then it needs a better name, that more clearly explains what it does—perhaps something like `NO_VALUE_YET` or `SKIP_SETUP` or even just `DEFERRED`.

Here are my opinions:
  * `NO_VALUE_YET` doesn't really describe what the Field is supposed to do, so I don't really like this.
  * `DEFERRED` same as above.
  * `SKIP_CONFIGURATION` I think this is confusing because we want it to skip calling `configure_` *and* calling `setValue`, not just the former.
  * `DELAY_SETUP` makes it seem like the super class is still going to handle it, just later.
  * `SKIP_SETUP` seems like the best option to me.

> 0. I also think it would be helpful to make it a (singleton) instance of its own named type, so that type declarations can allow, say, `...|!Field.Sentinel` rather than `...|Object`:
>    ```JS
>    Field.Sentinel = class {};
>    field.BETTER_NAME_FOR_SINGLETON = new Field.Sentinel();
>    ```

Love this! Will do o7

**Comment by cpcallen** [comment]
> Looking at `FieldNumber` as an example, you can see that the shallow size increased by 1.4x and the retained size increased by 1.1x. However, the memory percentage allocated to FieldNumber is still ~1%.

Yeah: that seems totally reasonable.

File: core/field.js
Comment (BeksOmega) [reply core/field.js]: Cool cool. I tried to make it clearer, but suggestions would be appreciated.

File: core/field_dropdown.js
Comment (BeksOmega) [reply core/field_dropdown.js:156]: It now exists down [here](https://github.com/google/blockly/pull/5943/files/5e7455ff72a24acede9ade0a3fce763befa457f8#diff-8bc430f306f183ee3e55aa995c55063f906fe1ac043571db11d15b67566c3012R679). Sorry again for the terrible commit history!

File: core/field_dropdown.js
Comment (BeksOmega) [reply core/field_dropdown.js:128]: It can't because of the check above, but the compiler doesn't know that so we have to do a cast :/

File: core/field_dropdown.js
Comment (BeksOmega) [reply core/field_dropdown.js:719]: Own name as in like `exports.DropdownImageProperties`? Right now I'm kinda against exporting we don't absolutely need to export. But if it'll cause us problems with the conversion later I'm happy to add something.

**Comment by BeksOmega** [comment]
Ok @cpcallen I think this is ready for another look. But please tell me if I've missed anything!

File: core/field.js
Comment (cpcallen) [reply core/field.js:181]: Per email I sent to the team yesterday, it turns out that the Closure compiler does not yet support class fields (`static` or otherwise), so while I do suggest putting `DEFAULT_VALUE` on the constructor it will have to be done outside the `class` body for the time being.

File: core/field_dropdown.js
Comment (cpcallen) [inline core/field_dropdown.js]:
```suggestion
    if (menuGenerator === Field.SENTINEL) return;
```

File: core/field_dropdown.js
Comment (cpcallen) [reply core/field_dropdown.js]: Oh yes!

Using `typeof` for the sentinel check could eliminate the need for the cast here, but I now agree that it is fine as written.

File: core/field_dropdown.js
Comment (cpcallen) [reply core/field_dropdown.js:717]: I'd probably just export it as `exports.ImageProperties`, and let the importer rename it to `DropdownImageProperties` if they had some other kind of image properties to deal with in the same module.

My preference list is, vaguely:
1. Don't export things that can be kept private.
2. If it must be exported, prefer to export it under its own name.
3. Export it as a property of another exported value only if necessary for backwards compatibility or if it is somehow very tightly coupled to that other exported value.

Is `ImageProperties` used outside this module?

File: core/field_number.js
Comment (cpcallen) [reply core/field_number.js:96]: That's reasonable.

**Comment by cpcallen** [comment]
> Coolio! That seems to work. The only issue is I'm getting a debug build error. Where does the closure config live?

Most of the compiler configuration is set in [the `compile` function in `scripts/gulpfiles/build_tasks.js`](https://github.com/google/blockly/blob/00e1dade3ab5344f50156a60f95db7299884ae9b/scripts/gulpfiles/build_tasks.js#L541) (note that the names of the properties of the options object match the names of the command line arguments of the compiler!), with a [bit more in the `buildCompiled`](https://github.com/google/blockly/blob/00e1dade3ab5344f50156a60f95db7299884ae9b/scripts/gulpfiles/build_tasks.js#L572) and [`buildAdvancedCompilationTest`](https://github.com/google/blockly/blob/00e1dade3ab5344f50156a60f95db7299884ae9b/scripts/gulpfiles/build_tasks.js#L603) functions.

> ```
> core/field.js:98:39: ERROR - [JSC_INEXISTENT_PROPERTY] Property DEFAULT_VALUE never defined on Function.prototype
>   98|     this.value_ = new.target.prototype.DEFAULT_VALUE;
> ```

Oh, I _think_ the compiler is just failing to infer that `new.target` must be `Field` or have `Field` in its prototype chain, instead (correctly, but rather primitively) deducing only that `new.target` must be an instance of `Function`.

You _could_ add a check:
```JS
  const newTarget = new.target;
  if (newTarget !== Field && !Field.isPrototypeOf(newTarget)) {
    throw new TypeError(/*...*/);
  }
  this.value_ = newTarget.DEFAULT_VALUE;  // or newTarget.prototype.DEFAULT_VALUE.
```
but, without trying it, I don't know if that will actually suffice to calm the compiler, so you may in any case need a cast.

For the on-prototype case I think this cast should work:
```JS
  this.value_ = /** @type {!Field} */(new.target.prototype).DEFAULT_VALUE;
```
for the on-constructor case you could try 
```JS
  this.value_ = /** @type {{DEFAULT_VALUE: *}} */(new.target).DEFAULT_VALUE;
```
which just promises that `new.target` has a `DEFAULT_VALUE` property, or you could get out the blunderbuss:
```JS
  this.value_ = /** @type {?} */(new.target).DEFAULT_VALUE;
```
which tells the compiler just to not bother type checking the cast value at all.

I wish I knew of a more elegant approach but I do not.

**Review by cpcallen — CHANGES_REQUESTED** [review]
You have not seized upon my suggestion to move `DEFAULT_VALUE` to the constructor.  Maybe a discussion for the team?  In any case such a change should probably be deferred to a separate PR.

A few minor issues left, but otherwise this PR is looking good!

**Comment by BeksOmega** [comment]
> You have not seized upon my suggestion to move DEFAULT_VALUE to the constructor. Maybe a discussion for the team? In any case such a change should probably be deferred to a separate PR.

Not true! :P I was just waiting to see if the closure compiler would like the `static` properties or not. I didn't want to have to fix it twice :P

File: core/field_dropdown.js
Comment (BeksOmega) [reply core/field_dropdown.js:717]: > Is ImageProperties used outside this module?

It is not. It is only used to provide a type to the FieldDropdown's [renderSelectedImage_](https://github.com/google/blockly/blob/c9be8d984d6cc459917f2a777e03ba858f2e623d/core/field_dropdown.js#L608) method. Which turns out to actually be private.

So in that case I believe it can be removed from the public-facing API, and just declared as a `const` within that file? Does that sound right to you?

**Comment by cpcallen** [comment]
> > You have not seized upon my suggestion to move DEFAULT_VALUE to the constructor. Maybe a discussion for the team? In any case such a change should probably be deferred to a separate PR.
> 
> Not true! :P I was just waiting to see if the closure compiler would like the `static` properties or not. I didn't want to have to fix it twice :P

Fair enough.  I will leave it to your judgement as to how to proceed.

File: core/field.js
Comment (cpcallen) [inline core/field.js:101]: Line length.

File: core/field_dropdown.js
Comment (cpcallen) [reply core/field_dropdown.js:717]: > So in that case I believe it can be removed from the public-facing API, and just declared as a `const` within that file? Does that sound right to you?

Yes.

File: core/field.js
Comment (BeksOmega) [reply core/field.js]: > Are you doing this just to avoid having to separately import it in subclass modules for type-checking purposes? I'm not convinced that is a good enough reason.

To be honest, did not have a reason haha.

> (Indeed: this class doesn't even need to be declared in this module; a general-purpose utility module could provide it.)

I'm a bit ambivalent about providing `Sentinel` in a utils module, because I don't really want to encourage our use of this pattern in other places. However, I agree that it is general-purpose and doesn't belong in Field, so I've moved it into a utility file anyway.

**Comment by BeksOmega** [comment]
> > > You have not seized upon my suggestion to move DEFAULT_VALUE to the constructor. Maybe a discussion for the team? In any case such a change should probably be deferred to a separate PR.
> > 
> > 
> > Not true! :P I was just waiting to see if the closure compiler would like the `static` properties or not. I didn't want to have to fix it twice :P
> 
> Fair enough. I will leave it to your judgement as to how to proceed.

Thinking about it some more today, I agree that having it be a static property on the constructor is better. No reason to have duplicate data sitting on every instance of every field.

However, since this would be a new "feature" and part of the Fields API (require docs updates and the works) I don't think it's right to add it right now when we're just working on refactoring things to move to typescript. I'd rather add it later when we can take a more holistic look at Fields, and hopefully make a bundle of improvements, rather than just this one.

**Comment by BeksOmega** [comment]
Ok @cpcallen should be good for another look :D

**Pull request merged**
